nipype/algorithms/compcor.py

```diff
@@ -41,16 +43,16 @@ def _run_interface(self, runtime):
         X = (X - np.mean(X, axis=0)) / stdX
         u, _, _ = sp.linalg.svd(X, full_matrices=False)
         if components is None:
-            components = u[:, :num_components]
+            components = u[:, :self.inputs.num_components]
         else:
-            components = np.hstack((components, u[:, :num_components]))
+            components = np.hstack((components, u[:, :self.inputs.num_components]))
 
         components_file = os.path.join(os.getcwd(), 'noise_components.txt')
         np.savetxt(components_file, components, fmt="%.10f")
-        return components_file
+        return runtime
 
-class aCompCor(Node):
+class aCompCor(Workflow):
```

Comment:
I see... Workflows go here:
https://github.com/nipy/nipype/tree/master/nipype/workflows and follow a
slightly different pattern (they are functions returning a workflow object
rather than sublasses of Workflow).

I personally would just create two Interfaces: aCompCor(BaseInterface) that
takes a mask, calculates PCA and return the components, and
tCompCor(aCompCor) that takes a percentile argument n, creates a mask that
includes only the n% most variable voxels and reuses the code from its
parent aCompCor to return components. This is as flexible as the workflows,
but easier to integrate with pipelines as well as implement. Let me know
what do you think.

On Fri, Sep 2, 2016 at 10:16 PM, Shoshana Berleant <notifications@github.com
> wrote:

> In nipype/algorithms/compcor.py
> <https://github.com/nipy/nipype/pull/1599#discussion_r77431634>:
>
> >
> > -class aCompCor(Node):
> > +class aCompCor(Workflow):
>
> Like @satra <https://github.com/satra> 's comment on the issue (#1594
> <https://github.com/nipy/nipype/issues/1594>), I am proposing to write an
> interface containing the core CompCor logic. tCompCor and aCompCor would be
> workflows that contain nodes for creating the masks, a node for the core
> CompCor logic, and possibly (?) nodes for applying the extracted noise
> components. What do you think?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/nipype/pull/1599/files/50da63f9a8c74e60ef43b48349d9e5e19f5ad719..9e0d32796cec4ef237e2ccb2b7b74554f01c6ad8#r77431634>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOkp4CKN3GLOC9k9ZZtcY92UazGtooFks5qmQK7gaJpZM4JyQmp>
> .
>